src/dissect/dissect.c

```diff
@@ -207,139 +273,438 @@ static int parse_argv(int argc, char *argv[]) {
         case ACTION_MOUNT:
                 if (optind + 2 != argc)
                         return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
-                                               "Expected a file path and mount point path as only arguments.");
+                                               "Expected an image file path and mount point path as only arguments.");
 
                 arg_image = argv[optind];
                 arg_path = argv[optind + 1];
                 break;
 
+        case ACTION_EXTRACT:
+                if (argc < optind + 2 || argc > optind + 3)
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                               "Expected an image file path, a source path and an optional destination path as only arguments.");
+
+                arg_image = argv[optind];
+                arg_source = argv[optind + 1];
+                arg_target = argc > optind + 2 ? argv[optind + 2] : "-" /* this means stdout */ ;
+
+                arg_flags |= DISSECT_IMAGE_READ_ONLY;
+                break;
+
+        case ACTION_ADD:
+                if (argc < optind + 2 || argc > optind + 3)
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                               "Expected an image file path, an optional source path and a destination path as only arguments.");
+
+                arg_image = argv[optind];
+
+                if (argc > optind + 2) {
+                        arg_source = argv[optind + 1];
+                        arg_target = argv[optind + 2];
+                } else {
+                        arg_source = "-"; /* this means stdin */
+                        arg_target = argv[optind + 1];
+                }
+
+                break;
+
         default:
                 assert_not_reached("Unknown action.");
         }
 
         return 1;
 }
 
-static int run(int argc, char *argv[]) {
-        _cleanup_(loop_device_unrefp) LoopDevice *d = NULL;
+static int action_dissect(DissectedImage *m, LoopDevice *d) {
+        _cleanup_(table_unrefp) Table *t = NULL;
+        uint64_t size;
+        int r;
+
+        assert(m);
+        assert(d);
+
+        printf("      Name: %s\n", basename(arg_image));
```

Comment:
indentation looks a bit strange with name/size more indented than the rest?

```
$ sudo buildx86/systemd-dissect img.raw
      Name: img.raw
      Size: 4.0K

Machine ID: 3d8a4aab19c64398aa48782f0a5e4ce8
OS Release: PRETTY_NAME=Debian GNU/Linux 10 (buster)
            NAME=Debian GNU/Linux
            VERSION_ID=10
            VERSION=10 (buster)
            VERSION_CODENAME=buster
            ID=debian
            HOME_URL=https://www.debian.org/
            SUPPORT_URL=https://www.debian.org/support
            BUG_REPORT_URL=https://bugs.debian.org/

RW DESIGNATOR PARTITION UUID FSTYPE   ARCHITECTURE VERITY   NODE    PARTNO
ro root       -              squashfs -            external img.raw      -
```

Reply:
what do you mean? everything above the partition table  is aligned along the `:`, like we usually do it.

Reply:
Ah I see - I guess my brain was expecting to see Name/Size left-aligned. If that's how all the tables are printed then it's fine.